plugins/metrics/pkg/cortex/intercept.go

```diff
@@ -0,0 +1,192 @@
+
+type RequestInterceptor interface {
+	Intercept(ctx context.Context, req *cortexpb.WriteRequest, handler WriteHandlerFunc) (*cortexpb.WriteResponse, error)
+}
+
+type federatingInterceptor struct {
+	conf InterceptorConfig
+
+	mdTracker *metadataTracker
+}
+
+type InterceptorConfig struct {
+	// The label to use as the tenant ID
+	IdLabelName string
+	// Whether to drop the tenant ID label in the output
+	DropIdLabel bool
+}
+
+func NewFederatingInterceptor(conf InterceptorConfig) RequestInterceptor {
+	return federatingInterceptor{
+		conf:      conf,
+		mdTracker: newMetadataTracker(),
+	}
+}
+
+// Intercept takes a write request and splits it into multiple requests, grouped
+// by tenant id found within the configured label name.
+// The outgoingCtx must have an org id set, as well as outgoing metadata, to
+// use as the default tenant id for timeseries without the configured label.
+func (t federatingInterceptor) Intercept(
+	outgoingCtx context.Context, req *cortexpb.WriteRequest,
+	handler WriteHandlerFunc,
+) (*cortexpb.WriteResponse, error) {
+	if len(req.Metadata) > 0 {
+		// Metadata comes in periodically, separate from timeseries, and in one
+		// large batch. Because we are splitting timeseries samples to other
+		// tenants, we also need to replicate the metadata to those tenants.
+		// We need to keep track of names of metrics for which we have sent
+		// timeseries data to, so that we can send metadata for those metrics
+		// to the same tenants the next time metadata comes in.
+		var errs []error
+		_, err := handler(outgoingCtx, req)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		errs = append(errs, t.mdTracker.replicate(req.Metadata, func(tenantId string, metricMetadata []*cortexpb.MetricMetadata) error {
+			req.Metadata = metricMetadata
+			md, _, _ := metadata.FromOutgoingContextRaw(outgoingCtx)
+			md[lowerOrgIDHeaderName] = []string{tenantId}
+			if _, err := handler(outgoingCtx, req); err != nil {
+				return err
+			}
+			return nil
+		})...)
+		return nil, errors.Join(errs...)
+	}
+	// for each time series in the request, check if there is a matching tenant ID label.
+	// if there is, start a new request for this and all subsequent entries with that tenant ID.
+	defaultId, err := user.ExtractOrgID(outgoingCtx)
+	if err != nil {
+		panic(err)
+	}
+	var buckets = map[string][]cortexpb.PreallocTimeseries{
+		defaultId: {},
+	}
+TIMESERIES:
+	for _, ts := range req.Timeseries {
+		nameLabelIndex := -1
+		for i, l := range ts.Labels {
+			// in practice, this appears to always be index 0
+			if l.Name == "__name__" {
+				nameLabelIndex = i
+				break
+			}
+		}
+		if nameLabelIndex == -1 {
+			continue
+		}
+		for i, l := range ts.Labels {
+			if l.Name == t.conf.IdLabelName {
+				if t.conf.DropIdLabel {
+					ts.Labels = append(ts.Labels[:i], ts.Labels[i+1:]...)
+				}
+				buckets[l.Value] = append(buckets[l.Value], ts)
+				t.mdTracker.track(ts.Labels[nameLabelIndex].Value, l.Value)
+				continue TIMESERIES
```

Comment:
this seems potentially expensive in the worst case (high volume of metrics, high number of labels, many drops)

Reply:
that scenario may never realistically happen, but maybe it makes sense to eventually add metrics for the federating interceptor itself, kinda like prometheus's scrape metrics

Reply:
I can try to optimize this a bit.

Reply:
Working on optimizing this significantly, since it is in the hot path -- making good progress.